spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPartitioningAwareScan.java

```diff
@@ -0,0 +1,244 @@
+  private List<ScanTaskGroup<T>> taskGroups = null; // lazy cache of task groups
+  private StructType groupingKeyType = null; // lazy cache of the grouping key type
+  private StructLikeSet groupingKeys = null; // lazy cache of grouping keys
+
+  SparkPartitioningAwareScan(
+      SparkSession spark,
+      Table table,
+      Scan<?, ? extends ScanTask, ? extends ScanTaskGroup<?>> scan,
+      SparkReadConf readConf,
+      Schema expectedSchema,
+      List<Expression> filters) {
+
+    super(spark, table, readConf, expectedSchema, filters);
+
+    this.scan = scan;
+    this.preserveDataGrouping = readConf.preserveDataGrouping();
+
+    if (scan == null) {
+      this.specs = Collections.emptySet();
+      this.tasks = Collections.emptyList();
+      this.taskGroups = Collections.emptyList();
+    }
+  }
+
+  protected abstract Class<T> taskJavaClass();
+
+  protected Scan<?, ? extends ScanTask, ? extends ScanTaskGroup<?>> scan() {
+    return scan;
+  }
+
+  @Override
+  public Partitioning outputPartitioning() {
+    Preconditions.checkState(taskGroups() != null, "Task groups must be planned");
+
+    if (groupingKeyType().fields().isEmpty()) {
+      return new UnknownPartitioning(taskGroups().size());
+    } else {
+      return new KeyGroupedPartitioning(groupingKeyTransforms(), taskGroups().size());
+    }
+  }
+
+  @Override
+  protected StructType groupingKeyType() {
+    if (groupingKeyType == null) {
+      if (preserveDataGrouping) {
+        this.groupingKeyType =
+            org.apache.iceberg.Partitioning.groupingKeyType(expectedSchema(), specs());
+      } else {
+        this.groupingKeyType = StructType.of();
+      }
+    }
+
+    return groupingKeyType;
+  }
+
+  private Transform[] groupingKeyTransforms() {
+    Set<Integer> groupingKeyFieldIds =
+        groupingKeyType().fields().stream()
+            .map(Types.NestedField::fieldId)
+            .collect(Collectors.toSet());
+
+    List<PartitionField> groupingKeyFields = Lists.newArrayList();
+
+    for (PartitionSpec spec : specs()) {
+      for (PartitionField field : spec.fields()) {
+        if (groupingKeyFieldIds.contains(field.fieldId())) {
+          groupingKeyFields.add(field);
+          groupingKeyFieldIds.remove(field.fieldId());
+        }
+      }
+    }
+
+    return Spark3Util.toTransforms(table().schema(), groupingKeyFields);
+  }
+
+  protected Set<PartitionSpec> specs() {
+    if (specs == null) {
+      Set<PartitionSpec> taskSpecs = Sets.newHashSet();
+      for (T task : tasks()) {
+        taskSpecs.add(task.spec());
+      }
+      this.specs = taskSpecs;
+    }
+
+    return specs;
+  }
+
+  @SuppressWarnings("unchecked")
+  protected synchronized List<T> tasks() {
+    if (tasks == null) {
+      try (CloseableIterable<? extends ScanTask> taskIterable = scan.planFiles()) {
+        List<T> plannedTasks = Lists.newArrayList();
+
+        for (ScanTask task : taskIterable) {
+          ValidationException.check(
```

Comment:
is this desirable to do the validation in the for LOOP? the exception will be thrown anyway.

Reply:
This check is to provide a reasonable exception if there is a task of an unsupported type. Since we may have a combination of different tasks, each task is being validated.

I don't think it would be super expensive to perform this validation but it is kind of optional too.